src/hotspot/share/gc/g1/g1ConcurrentMark.hpp

```diff
@@ -136,10 +136,92 @@ class G1CMMarkStack {
     G1TaskQueueEntry data[EntriesPerChunk];
   };
 
-  size_t _max_chunk_capacity;    // Maximum number of TaskQueueEntryChunk elements on the stack.
+  class ChunkAllocator {
+    // The chunk allocator relies on a growable array data structure that allows resizing without the
+    // need to copy existing items. The fundamental approach involves organizing the array into chunks,
+    // essentially creating an "array of arrays"; referred to as buckets in this implementation. To
+    // facilitate efficient indexing, the size of the first chunk is set to a power of 2. This choice
+    // allows for quick conversion of an array index into a bucket index and the corresponding offset
+    // within the bucket. Additionally, each new chunk added to the growable array doubles the capacity of
+    // the growable array.
+    //
+    // Illustration of the Growable Array data structure.
+    //
+    //        +----+        +----+----+
+    //        |    |------->|    |    |
+    //        +----+        +----+----+
+    //        |    |        +----+----+
+    //        |    |------->|    |    |
+    //        +----+        +----+----+
+    //        |    |        +-----+-----+-----+-----+
+    //        |    |------->|     |     |     |     |
+    //        +----+        +-----+-----+-----+-----+
+    //        |    |        +-----+-----+-----+-----+-----+-----+-----+----+
+    //        |    |------->|     |     |     |     |     |     |     |    |
+    //        +----+        +-----+-----+-----+-----+-----+-----+-----+----+
+    //
```

Comment:
"... doubles the capacity of the growable array" this is the general approach to growing "vectors" / "arrays" so what I was aiming for is what you state in the second paragraph as "each new bucket doubles the capacity of this chunk-allocator".

What happens if all buckets have the same capacity?
Then that is a different approach, requires  deciding on the fixed "bucket capacity" rather than "growing out" the buckets.

Reply:
> ...  is the general approach to growing "vectors" / "arrays"

That's true, but in this case, there is a 'ceiling' value. The overall goal mark-stack-resizing is to strike a balance between two conflicting constraints: minimizing the number of expands and minimizing the average mark-stack footprint. The approach I have in mind is to 'hardcode' the #expands to a constant, let's say 4. This way, each expand accounts for (max - init) / 4. (Using 2 seems too small, and the next even number is 4.)

For example, when init = 4 and max = 8 (in whatever suitable unit), the mark-stack capacity can grow like this:

double-on-expand: 4 -> 8
equal-delta ((8-4)/4 == 1): 4 -> 5 -> 6 -> 7 -> 8

In the double-on-expand case, there is only a single expand, effectively losing the elasticity of a growable mark-stack. Also, I think equal-delta will have a simpler implementation.

Reply:
>  The approach I have in mind is to 'hardcode' the #expands to a constant, let's say 4.
I don't like the hardcoded approach because we end up picking random numbers such as the 4 expands in your example without much reason to support that.

> In the double-on-expand case, there is only a single expand, effectively losing the elasticity of a growable mark-stack.
I wanted to maintain the expansion by doubling capacity as is in the current implementation.  Any attempts to optimise seem premature to me.

> equal-delta ((8-4)/4 == 1): 4 -> 5 -> 6 -> 7 -> 8
You are growing too slow, if the initial 4 overflows, and you grow by 1, with high probability you will overflow the 5, actually thought you would suggest growing by the init (4). But again, I don't like the idea of hardcoded number of expands as I am not sure about the motivation behind it.

>  I think equal-delta will have a simpler implementation.
I find the proposed approach really easy.